protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -81,148 +51,240 @@ where
+                    let nonce = self.rng.gen();
+                    self.pending_requests.push_back(DialRequest {
+                        addrs: vec![addr.clone()],
+                        nonce,
+                    });
+                    self.pending_nonces.insert(nonce);
+                }
+            }
+            FromSwarm::ConnectionEstablished(ConnectionEstablished {
+                peer_id,
+                connection_id,
+                ..
+            }) => {
+                self.peers_to_handlers
+                    .entry(peer_id)
+                    .or_insert(connection_id);
+            }
+            FromSwarm::ConnectionClosed(ConnectionClosed {
+                peer_id,
+                connection_id,
+                ..
+            }) => {
+                tracing::trace!("connection with {peer_id:?} closed");
+                self.handle_no_connection(peer_id, connection_id);
+            }
+            FromSwarm::DialFailure(DialFailure {
+                peer_id: Some(peer_id),
+                error,
+                connection_id,
+            }) => {
+                tracing::trace!("dialing {peer_id:?} failed: {error:?}");
+                self.handle_no_connection(peer_id, connection_id);
+            }
+            _ => {}
+        }
     }
 
     fn on_connection_handler_event(
         &mut self,
         peer_id: PeerId,
-        _connection_id: ConnectionId,
-        event: <Self::ConnectionHandler as ConnectionHandler>::ToBehaviour,
+        connection_id: ConnectionId,
+        event: <Handler as ConnectionHandler>::ToBehaviour,
     ) {
+        self.peers_to_handlers
+            .entry(peer_id)
+            .or_insert(connection_id);
         match event {
-            ToBehaviour::ResponseInfo(response_info) => {
-                let event = Event::CompletedReachabilityTest(
-                    self.handle_response_info(peer_id, response_info),
+            Either::Right(Ok(nonce)) => {
+                if self.pending_nonces.remove(&nonce) {
+                    tracing::trace!("Received pending nonce from {peer_id:?}");
+                } else {
+                    tracing::warn!("Received unexpected nonce from {peer_id:?}, this means that another node tried to be reachable on an address this node is reachable on.");
+                }
+            }
+            Either::Right(Err(err)) => {
+                tracing::debug!("Dial back failed: {:?}", err);
+            }
+            Either::Left(RequestToBehaviour::PeerHasServerSupport) => {
+                if !self.known_servers.contains(&peer_id) {
+                    self.known_servers.push(peer_id);
+                }
+            }
+            Either::Left(RequestToBehaviour::TestCompleted(Ok(TestEnd {
+                dial_request: DialRequest { nonce, addrs },
+                suspicious_addr,
+                reachable_addr,
+            }))) => {
+                if self.pending_nonces.remove(&nonce) {
```

Comment:
Isn't this always going to be false because you've removed the nonce already in line 135?

Reply:
Not necessarily. 
It's possible that the server thinks it successfully dialed back, but we never got that dial back. A possible example could be a badly implemented server. If we don't check that, what's the point in even keeping track of the nonces and the dial backs?

Reply:
I lost track of what that was, but I think it's solved, since it works.

Reply:
The problem is that because you removed the received nonce further above, you cannot differentiate between the following states here:

- We received the correct nonce from the server
- We received a nonce that we've never heard of

Instead, I think we should:

- "Mark" a nonce as successfully received when we get it from the `dial_back::Handler`
- Ensure that once we receive the response from the server that we did in fact mark the associated nonce as "successful". If we didn't we should throw away the response from the server.

Reply:
Got it, changed it.